### PR TITLE
Improvement: DO-3044: Performance improvement and refactor of all list components

### PR DESCRIPTION
**Pull request by Roman-Kornev**
## Motivation and Context

It was reported that scrolling in our list components feel sluggish. We recently did a performance pass that addressed scrolling when the list was NOT open with significant boosts across the board.
This PR improves performance of scrolling the actual list by 485x for big lists (10000 elements) for all list components:
- Select
- Multiselect
- ComboBox
- SectionedList
- DatepickerSelect
- ContextMenu

To optimize this even further we can use virtualized rendering to only render subset of the items dynamically, but at this point the performance feels acceptable.

Fixes https://github.com/causalens/dara/issues/307.

## Implementation Description

- Every single component was rendering items slightly differently. Refactored it all into a single memoized `DropdownList` component, which now uses `ListItem` as a default renderer, but can be customized with a render prop.
- The components were also rerendering due to setting their `hover` status with JS instead of CSS. All references to `highlightedIndex` were replaced with a `: hover` style.
- In some cases the chevron was rerendering as well - it has been extraced and memoized now.

All components are now memoized properly and won't rerender when scrolled.

## How Has This Been Tested?

React profiler + chrome profiler. Didn't notice any significant memory impact.

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [x] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

### BEFORE, a frame takes 193ms:
image

https://github.com/causalens/dara-ui/assets/119343173/17037e59-8dff-4296-8996-a647b47a1b37

Notice how every single item in the list is rerendered on scroll.

### AFTER, a frame takes 0.2ms:
image

https://github.com/causalens/dara-ui/assets/119343173/6b58554d-037a-437b-b82f-d246a9a54be4

### CODE REVIEW

File: packages/ui-components/src/shared/list-item.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    return (
+        <StyledListItem
+            {...itemProps}
+            className={itemClass ? `${itemClassName as string} ${itemClass}` : (itemClassName as string)}
+            title={title}
+            size={size}
+            item={item}
+        >
+            {' '}
```
Comment (krzysztof-causalens) [inline packages/ui-components/src/shared/list-item.tsx:93]: Is this intentional?

File: packages/ui-components/src/utils/chevron.tsx [hunk 2]
```diff
@@ -51,18 +51,16 @@ const Icon = styled(ChevronDown).withConfig({ shouldForwardProp })<ChevronProps>
...
+        disabled={props.disabled}
+        isOpen={props.isOpen}
+        onClick={props.onClick}
+        ref={ref}
+        style={props.style}
+    />
+));
 
-export default React.forwardRef(Chevron);
+export default Chevron;
```
Comment (krzysztof-causalens) [inline packages/ui-components/src/utils/chevron.tsx:66]: Should this have a memo as well?

**Review by krzysztof-causalens — APPROVED** [review]
LGTM, nice job

File: packages/ui-components/src/utils/chevron.tsx (hunk 2 above)
Comment (Roman-Kornev) [reply packages/ui-components/src/utils/chevron.tsx]: The button itself is already memo'ed, but I guess it doesn't hurt if any other component uses it directly.

**Review by patricia-causalens — CHANGES_REQUESTED** [review]
Hi Roman, something seems to have gone wrong with the styling of the select in the datepicker, the items now have a weird space between them

Screenshot 2024-06-03 at 09 15 44

**Comment by Roman-Kornev** [comment]
@patricia-causalens Thanks for checking it out. Turns out the custom styling was not applied properly when using higher-order styled components. Given that the datepicker select looks distinct enough from a regular select item, it now renders the styled version directly instead of inheriting from the generic list item.

A couple of other issues came up:
- The scroll-into-view functionality of downshift doesn't work if the item is memoized. This item is now manually de-memoized when the menu is opened to let downshift automatically scroll it into view. Profiling this new change did not highlight any performance issues.
- Keyboard navigation highlighting. Replacing all highlightedIndex styling with CSS `:hover` turned out to cause issues for highlighting items when using the keyboard, so a similar fix needed to be applied. In order to prevent rerenders when the mouse is moved, a new state is hooked directly into downshift's `onHighlightedIndexChange` that updates only when using the keyboard and resets when using the mouse. This results in a smooth experience when using the mouse as no rerenders are happening (only CSS `:hover` is used) and only the 2 affected items are rerendered when using the keyboard.

**Review by patricia-causalens — CHANGES_REQUESTED** [review]
Good catch on the scroll! Nice work! Had another play and found a few more issues:

1. The currently selected doesn't seem to be highlighted anymore when opening the select:
Before:
Screenshot 2024-06-03 at 20 37 59

Now:
Screenshot 2024-06-03 at 20 38 19

2. The select in the datepicker within the filter currently selected appears with white text and without the background color:
Screenshot 2024-06-03 at 20 36 48

3. The font and overall dropdown of the select is also still a bit off for the datepicker, it should have 0.75rem font and the length should be such that it covers the numbers of the datepicker behind it:
Screenshot 2024-06-03 at 20 35 24

I will slack you the figma with all the designs for reference :)

**Comment by Roman-Kornev** [comment]
Turns out there was a global style being applied to only that particular datepicker filter which was overriding some colors. Since dropdowns can now customize how elements are rendered, that global style was removed.

The dropdowns are now aligned with the design:
- Currently selected element now has a different color from the highlighted element, so it's easier to tell which one is which when hovering
- The datepicker is now correctly taking into account the gap between elements in its height calculation
- Datepicker's items now have rounded corners

**Review by patricia-causalens — APPROVED** [review]
LGTM, thanks for fixing the different stylings 😄
